compiler_opt/es/blackbox_learner.py

```diff
@@ -0,0 +1,319 @@
+      tf.summary.scalar(
+          'reward/regression_probability_train',
+          len(train_regressions) / len(rewards),
+          step=self._step)
+
+      tf.summary.scalar(
+          'reward/regression_avg_train',
+          np.mean(train_regressions),
+          step=self._step)
+
+      # The "max regression" is the min value, as the regressions are negative.
+      tf.summary.scalar(
+          'reward/regression_max_train',
+          min(train_regressions, default=0),
+          step=self._step)
+
+      train_wins = [reward for reward in rewards if reward > 0]
+      tf.summary.scalar(
+          'reward/win_probability_train',
+          len(train_wins) / len(rewards),
+          step=self._step)
+
+  def _save_model(self) -> None:
+    """Save the model."""
+    logging.info('Saving the model.')
+    self._policy_saver_fn(
+        parameters=self._model_weights, policy_name=f'iteration{self._step}')
+
+  def get_model_weights(self) -> npt.NDArray[np.float32]:
+    return self._model_weights
+
+  def _get_results(
+      self, pool: FixedWorkerPool,
+      perturbations: List[bytes]) -> List[concurrent.futures.Future]:
+    if not self._samples:
+      for _ in range(self._config.total_num_perturbations):
+        sample = self._sampler.sample(self._config.num_ir_repeats_within_worker)
+        self._samples.append(sample)
+        # add copy of sample for antithetic perturbation pair
+        if self._config.est_type == (
+            blackbox_optimizers.EstimatorType.ANTITHETIC):
+          self._samples.append(sample)
+
+    compile_args = zip(perturbations, self._samples)
+
+    _, futures = buffered_scheduler.schedule_on_worker_pool(
+        action=lambda w, v: w.temp_compile(v[0], v[1]),
+        jobs=compile_args,
+        worker_pool=pool)
+
+    not_done = futures
+    # wait for all futures to finish
+    while not_done:
+      # update lists as work gets done
+      _, not_done = concurrent.futures.wait(
+          not_done, return_when=concurrent.futures.FIRST_COMPLETED)
+
+    return futures
+
+  def _get_policy_as_bytes(self,
+                           perturbation: npt.NDArray[np.float32]) -> bytes:
+    sm = tf.saved_model.load(self._tf_policy_path)
+    # devectorize the perturbation
+    policy_utils.set_vectorized_parameters_for_policy(sm, perturbation)
+
+    with tempfile.TemporaryDirectory() as tmpdir:
+      sm_dir = os.path.join(tmpdir, 'sm')
+      tf.saved_model.save(sm, sm_dir, signatures=sm.signatures)
+      src = os.path.join(self._tf_policy_path, policy_saver.OUTPUT_SIGNATURE)
+      dst = os.path.join(sm_dir, policy_saver.OUTPUT_SIGNATURE)
+      tf.io.gfile.copy(src, dst)
+
+      # convert to tflite
+      tfl_dir = os.path.join(tmpdir, 'tfl')
+      policy_saver.convert_mlgo_model(sm_dir, tfl_dir)
+
+      # create and return policy
+      policy_obj = policy_saver.Policy.from_filesystem(tfl_dir)
+      return policy_obj.policy
+
+  def run_step(self, pool: FixedWorkerPool) -> None:
```

Comment:
do you need a docstring here?  at least mention that this will block until the step is complete.

Reply:
Do you mind clarifying what you mean by block? I am not sure what it will be blocking. thanks

Reply:
i mean that there's gonna be a bunch of I/O and executions running while this waits for the responses; so it doesn't instantaneously return.